pennylane/devices/qutrit_mixed/apply_operation.py

```diff
@@ -0,0 +1,201 @@
+):
+    """Apply and operator to a given state.
+
+    Args:
+        op (Operator): The operation to apply to ``state``
+        state (TensorLike): The starting state.
+        is_state_batched (bool): Boolean representing whether the state is batched or not
+        debugger (_Debugger): The debugger to use
+
+    Returns:
+        ndarray: output state
+
+    .. warning::
+
+        ``apply_operation`` is an internal function, and thus subject to change without a deprecation cycle.
+
+    .. warning::
+        ``apply_operation`` applies no validation to its inputs.
+
+        This function assumes that the wires of the operator correspond to indices
+        of the state. See :func:`~.map_wires` to convert operations to integer wire labels.
+
+        The shape of state should be ``[QUDIT_DIM]*(num_wires * 2)``, where ``QUDIT_DIM`` is
+        the dimension of the system.
+
+    This is a ``functools.singledispatch`` function, so additional specialized kernels
+    for specific operations can be registered like:
+
+    .. code-block:: python
+
+        @apply_operation.register
+        def _(op: type_op, state):
+            # custom op application method here
+
+    **Example:**
+
+    >>> state = np.zeros((3,3))
+    >>> state[0][0] = 1
+    >>> state
+    tensor([[1., 0., 0.],
+        [0., 0., 0.],
+        [0., 0., 0.]], requires_grad=True)
+    >>> apply_operation(qml.TShift(0), state)
+    tensor([[0., 0., 0.],
+        [0., 1., 0],
+        [0., 0., 0.],], requires_grad=True)
+
+    """
+    return _apply_operation_default(op, state, is_state_batched, debugger)
+
+
+def _apply_operation_default(op, state, is_state_batched, debugger):
+    """The default behaviour of apply_operation, accessed through the standard dispatch
+    of apply_operation, as well as conditionally in other dispatches.
+    """
+
+    return apply_operation_einsum(op, state, is_state_batched=is_state_batched)
+    # TODO add tensordot and benchmark for performance
+
+
+# TODO add diagonal for speed up.
+
+
+@apply_operation.register
+def apply_snapshot(op: qml.Snapshot, state, is_state_batched: bool = False, debugger=None):
+    """Take a snapshot of the mixed state"""
+    if debugger and debugger.active:
+        measurement = op.hyperparameters["measurement"]
+        if measurement:
+            # TODO replace with: measure once added
+            raise NotImplementedError  # TODO
+        if is_state_batched:
+            dim = int(np.sqrt(math.size(state[0])))
+            flat_shape = [math.shape(state)[0], dim, dim]
+        else:
+            dim = int(np.sqrt(math.size(state)))
+            flat_shape = [dim, dim]
+
+        snapshot = math.reshape(state, flat_shape)
+        if op.tag:
+            debugger.snapshots[op.tag] = snapshot
+        else:
+            debugger.snapshots[len(debugger.snapshots)] = snapshot
+    return state
+
+
+# TODO add special case speedups
+
+
+def _get_kraus(operation):  # pylint: disable=no-self-use
```

Comment:
Two comments about this function:
 - It gets called before it is defined, so I recommend moving it to the top of the file before the `apply_operation_einsum` function
 - At the moment it is only being called in one place; is there a particular reason for defining it separately, instead of just using an `if` statement? Is it going to be called by other, future functions?

Reply:
Yes it would be needed by tensordot if that was readded. should I leave it, (move to the top) or should I inline it and leave a comment?

Reply:
I suggest to inline and leave a comment.